feat!: add presentation frame

TimoGlastra opened the pull request:
Adds support for presentation frame in the `.present` method.

The logic in this PR already supports recursive disclosures in the presentation frame, so if #2 gets resolved, this could should not require any code changes.

There's two things that could probably be improved upon here:
- add more tests, especially for the lower level methods
- integrate the new method more with the existing code. I think this library will benefit from having some base util methods to traverse payloads/disclosures/etc...
- currently you need to pass the hasher method everywhere. I already added a memoization so that if a disclosure is hashed once, it won't re-hash every time, but  I think there will be benefit in create a disclosure type that stores the digest within the disclosure

Fixes #1

berendsliedrecht commented on src/sdJwt/disclosureMapping.ts:
```diff
@@ -0,0 +1,242 @@
+                const digest = item['...']
+                if (typeof digest !== 'string') {
+                    throw new SdJwtError(
+                        `Expected value of '...' to be of type string, but found ${typeof digest}`
```
```suggestion
                        `Expected value of key '...' to be of type string, but found ${typeof digest}`
```

Was confused for a second ;).

berendsliedrecht commented on src/sdJwt/disclosureMapping.ts:
```diff
@@ -0,0 +1,242 @@
+// This file contains helpers functions for mapping between disclosures entries and the payload of an SD-JWT.
```
I can do this later, but this would benefit from some comments :). Well, the whole repo could benefit from that..

TimoGlastra replied:
I'll add some before we merge. Better to do it now than later

berendsliedrecht commented on src/sdJwt/disclosures.ts:
```diff
@@ -6,6 +6,7 @@ export class Disclosure {
     private salt: string
     private key?: string
     private value: unknown
+    private _digest: string | undefined
```
```suggestion
    private digest?: string
```

Was never a fan of the `_` prefix.

TimoGlastra replied:
This will clash with the method with the same name

berendsliedrecht commented on src/sdJwt/presentationFrame.ts:
```diff
@@ -0,0 +1,93 @@
+        // If the value is false, it means we don't want to disclose it
+        if (node.value === false) continue
+
+        // FIXME: we should check whether the presentation frame matches the type of the payload. So if the presentation frame has an array as value, but the pretty payload
```
> So if the presentation frame has an array as value, but the pretty payload

What does this mean?

TimoGlastra replied:
Lets say the payload is

```
{
  "array": [1, 2]
}
```

But the presentation frame is:

```
{
  "array": { "firstKey" }
}
```

I'm nit really sure what will happen atm. Might be that it already handles this, but not sure

berendsliedrecht commented on src/sdJwt/sdJwt.ts:
```diff
@@ -305,89 +307,65 @@ export class SdJwt<
-        ) {
-            throw new SdJwtError('NaN and infinity are not allowed as indices')
-        }
+        // TODO: wouldn't it be easier if this returned the value so we don't have to use !
```
It could return `this` with the property defined or the value. Any preference?

TimoGlastra replied:
I think the value would be great.

Then it becomes:

```
const payload = this.assertPayload()
```

berendsliedrecht replied:
Yeah I think that is something I can add in a follow-up PR. Bothered me quite a bit as well.

berendsliedrecht commented on src/types/present.ts:
```diff
@@ -0,0 +1,15 @@
+export type PresentationFrame<T> = T extends Array<unknown>
```
Maybe there is a difference but can we not to `PresentationFrame = DisclosureFrame`

berendsliedrecht replied:
Maybe the disclosure frame can extend the presentation frame and add the decoy count on every level.

TimoGlastra replied:
I'll try to create a generic type, where you can pass extra properties.

However I noticed some issues with this type in general, that if you use the default record<string, unknown> it will not accept all types correctly

TimoGlastra replied:
Okay combined them 👍

berendsliedrecht commented on src/utils/utils.ts:
```diff
@@ -81,3 +86,62 @@ export const deleteByPath = (object: Record<string, unknown>, path: string) => {
+): boolean => {
+    try {
+        getByPath(item, path)
+        return true
```
I wish there was  built-in functionality to convert an error to boolean...

berendsliedrecht commented on tests/sdJwt.test.ts:
```diff
@@ -1263,7 +1218,10 @@ describe('sd-jwt', async () => {
             await rejects(
-                async () => await sdJwt.present([Infinity]),
+                async () =>
+                    await sdJwt.present({
```
Way better api :).

berendsliedrecht reviewed (approved):
Great changes! Some of the utility code is quite difficult to read, but indeed more tests will help with that. I can do that in the future and we can leave it for now like this.

TimoGlastra commented:
Okay addresses feedback, added some comments, added some extra tests and fixed some more bugs. I think this is now ready, unless any of the unresolved comments needs additional changes?
